File: 2-add-two-numbers/add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        res = ListNode(-1)
        tmp = res
        c = 0
        while l1 and l2:
            s = l1.val + l2.val + c
            c = s // 10
            tmp.next = ListNode(s % 10)
            tmp = tmp.next
            l1 = l1.next
            l2 = l2.next
        if l1:
            while l1:
                s = l1.val + c
                tmp.next = ListNode(s % 10)
                c = s // 10
                tmp = tmp.next
                l1 = l1.next
        if l2:
            while l2:
                s = l2.val + c
                tmp.next = ListNode(s % 10)
                c = s // 10
                tmp = tmp.next
                l2 = l2.next
        if c:
            tmp.next = ListNode(c)
        return res.next

File: 2-add-two-numbers/add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        int carry = 0;
        ListNode res = new ListNode();
        ListNode curRes = res;
        ListNode cur1 = l1;
        ListNode cur2 = l2;
        while(cur1 != null || cur2 != null) {
            int v1 = 0;
            int v2 = 0;
            if(cur1 != null) {
                v1 = cur1.val;
                cur1 = cur1.next;
            }
            if(cur2 != null) {
                v2 = cur2.val;
                cur2 = cur2.next;
            }
            int s = v1 + v2 + carry;
            ListNode newNode = new ListNode(s % 10);
            carry = s / 10;
            curRes.next = newNode;
            curRes = curRes.next;
        }
        if(carry > 0) {
            curRes.next = new ListNode(carry);
        }
        return res.next;
    }
}